util: Remove RandAddSeedPerfmon

hodlinator opened the pull request:
`RegQueryValueExA(HKEY_PERFORMANCE_DATA, ...)` sometimes hangs *bitcoind.exe* on Windows during startup, at least on CI.

We have other sources of entropy to seed randomness with on Windows, so should be alright removing this. Might resurrect if less drastic fix is found.

Hopefully sufficient to fix #30390.

CI debugged with temporary Windows stack trace dumping + Symbols in #30956.

hodlinator committed:
util: Remove RandAddSeedPerfmon

RegQueryValueExA(HKEY_PERFORMANCE_DATA, ...) sometimes hangs bitcoind.exe on Windows during startup, at least on CI.

We have other sources of entropy to seed randomness with on Windows, so should be alright removing this. Might resurrect if less drastic fix is found.

hodlinator commented:
For brief context, in 3 of 3 checked Windows stack dumps, the main callstack where it hangs has been:
```
 	ntdll.dll!NtReleaseSemaphore()	Unknown
 	KERNELBASE.dll!ReleaseSemaphore()	Unknown
 	WmiApRpl.dll!WmiReverseGuard::LeaveRead(long)	Unknown
 	WmiApRpl.dll!WmiReverseMemoryExt<class WmiReverseGuard>::Read(void *,unsigned long,unsigned long *,unsigned long)	Unknown
 	WmiApRpl.dll!WmiAdapterWrapper::GetValidity(unsigned long)	Unknown
 	WmiApRpl.dll!WmiAdapterWrapper::CollectObjects(unsigned short const *,void * *,unsigned long *,unsigned long *)	Unknown
 	WmiApRpl.dll!WmiAdapterWrapper::Collect(unsigned short const *,void * *,unsigned long *,unsigned long *)	Unknown
 	WmiApRpl.dll!WmiCollectPerfData(unsigned short const *,void * *,unsigned long *,unsigned long *)	Unknown
 	advapi32.dll!CallExtObj()	Unknown
 	advapi32.dll!QueryV1Provider()	Unknown
 	advapi32.dll!QueryExtensibleData()	Unknown
 	advapi32.dll!PerfRegQueryValueEx()	Unknown
 	advapi32.dll!PerfRegQueryValue()	Unknown
 	KERNELBASE.dll!BaseRegQueryValueInternal()	Unknown
>	KERNELBASE.dll!RegQueryValueExA()	Unknown
 	bitcoind.exe!`anonymous namespace'::RandAddSeedPerfmon(CSHA512 & hasher) Line 86	C++
 	bitcoind.exe!RandAddDynamicEnv(CSHA512 & hasher) Line 235	C++
 	bitcoind.exe!`anonymous namespace'::SeedStartup(CSHA512 & hasher, `anonymous-namespace'::RNGState & rng) Line 624	C++
 	bitcoind.exe!`anonymous namespace'::ProcRand(unsigned char * out, int num, `anonymous-namespace'::RNGLevel level, bool always_use_real_rng) Line 662	C++
 	[Inline Frame] bitcoind.exe!kernel::Context::{ctor}::__l2::<lambda_1>::operator()() Line 21	C++
 	[Inline Frame] bitcoind.exe!std::invoke(kernel::Context::{ctor}::__l2::<lambda_1> &&) Line 1704	C++
 	bitcoind.exe!std::call_once<`kernel::Context::Context'::`2'::<lambda_1>>(std::once_flag & _Once, kernel::Context::{ctor}::__l2::<lambda_1> && _Fx) Line 103	C++
 	bitcoind.exe!kernel::Context::Context() Line 23	C++
 	[Inline Frame] bitcoind.exe!std::make_unique() Line 3597	C++
 	bitcoind.exe!AppInit(node::NodeContext & node) Line 188	C++
 	bitcoind.exe!main(int argc, char * * argv) Line 277	C++
 	[Inline Frame] bitcoind.exe!invoke_main() Line 78	C++
 	bitcoind.exe!__scrt_common_main_seh() Line 288	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown
```
...suggesting some kind of (semaphore-related) deadlock sometimes occurs.

fanquake commented:
Concept ACK - can also remove `winreg.h`.

laanwj commented:
Concept ACK as explained in https://github.com/bitcoin/bitcoin/pull/30956#issuecomment-2425157662 - querying all performance counters has a potentially high overhead for the system.

hodlinator commented on src/randomenv.cpp:
```diff
@@ -227,8 +188,6 @@ void AddAllCPUID(CSHA512& hasher)
 
 void RandAddDynamicEnv(CSHA512& hasher)
 {
-    RandAddSeedPerfmon(hasher);
```
Adjusted in latest push (5cdca2f9d704f2f57ed49beaa764d31c8c55ca77), new comment:
```C++
// Dynamic environment data (clocks, resource usage, ...)
```

hodlinator commented:
Thanks! Incorporated feedback so far. Changed my personal master branch back to running a scheduled hourly Windows-only job with dump file generation, now including the fix from this PR: https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:master

hodlinator commented:
Updated 2 more comments along the lines of https://github.com/bitcoin/bitcoin/pull/31124#discussion_r1808780747 in latest push.

fanquake reviewed (approved):
ACK 9bb92c0e7ffe2426b4afb80ddd4b4c96968316b5

hebasto reviewed (approved):
ACK 9bb92c0e7ffe2426b4afb80ddd4b4c96968316b5, I have reviewed the code and it looks OK.

laanwj reviewed (approved):
Code review ACK  9bb92c0e7ffe2426b4afb80ddd4b4c96968316b5

achow101 commented:
ACK 9bb92c0e7ffe2426b4afb80ddd4b4c96968316b5

davidgumberg commented:
post-merge utACK https://github.com/bitcoin/bitcoin/pull/31124/commits/9bb92c0e7ffe2426b4afb80ddd4b4c96968316b5

laanwj commented:
> I haven't thought very seriously about what high quality entropy sources are and whether or not we have enough of them on Windows without this performance data

The high quality OS entropy source on windows is `CryptGenRandom`, and we use that.

Not opposed to collecting other ancillary data to mix in, though let's try to avoid something as crude as pulling data from all drivers and processes on the system.

hodlinator commented:
> The high quality OS entropy source on windows is `CryptGenRandom`, and we use that.

[A Microsoft article](https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom) from 2021 on `CryptGenRandom` states:
> **Important**  This API is deprecated. New and existing software should start using [Cryptography Next Generation APIs.](https://learn.microsoft.com/en-us/windows/desktop/SecCNG/cng-portal) Microsoft may remove this API in future releases.

So maybe we could create an issue to migrate to a more modern API?

laanwj commented:
Yes, it's deprecated, but IIRC the best compromise for the supported windows version. After #31172 we can switch to a whole slew of more modern APIs for windows.
